File: SupermarketSystem.Domain/Ordering/Order.cs
using SupermarketSystem.Domain.Common;

namespace SupermarketSystem.Domain.Ordering;

public enum OrderStatus
{
    Pending = 1,
    Accepted = 2,
    Completed = 3,
    Rejected = 4
}

/// <summary>
/// أساس ميزة "التطبيق للزبائن" (نقاش صاحب المشروع، مش مبنية عليه أي
/// شاشة تطبيق لسا - هذا الأساس بالباك إند بس). دورة حياة الطلب مقصودة
/// أربع حالات، مش اتنين:
///
///   Pending → Accepted → Completed
///           ↘ Rejected (بأي لحظة قبل Completed، بسبب إلزامي)
///
/// "Accepted" منفصلة عن "Completed" عمدًا: CompleteSaleCommand (فاتورة
/// البيع الحقيقية) بيتطلب سداد كامل فوري - ما بيسمح بفاتورة "غير
/// مدفوعة بعد". طلبات التوصيل عادة Cash on Delivery (الدفع وقت
/// التسليم)، فمافي مبلغ فعلي نقدر نسجّله كفاتورة حقيقية لحظة "الموافقة"
/// - بس لحظة "الإكمال" (لما الكاش يترجع فعليًا) مننشئ SaleInvoice حقيقية
/// (بإعادة استخدام CompleteSaleHandler نفسه، صفر تكرار منطق مخزون/دفع).
/// "Accepted" لسا مفيدة: بتعلّم الكاشير إنه التزم يجهّز الطلب.
///
/// لا تعديل على الأصناف بعد الإنشاء (قرار صاحب المشروع - تعديل أصناف
/// بيبطّئ العملية، فرق السعر/التوفر البسيط واقع مقبول، حل الحالات
/// الاستثنائية = رفض الطلب كامل بسبب واضح).
/// </summary>
public class Order : AuditableEntity, IBranchOwned
{
    public Guid CustomerId { get; private set; }
    public Guid BranchId { get; private set; }
    public OrderStatus Status { get; private set; }

    public string? DeliveryNote { get; private set; }
    public decimal? DeliveryLatitude { get; private set; }
    public decimal? DeliveryLongitude { get; private set; }

    public Guid? DecidedByUserId { get; private set; }
    public DateTime? DecidedAtUtc { get; private set; }
    public string? RejectionReason { get; private set; }

    public Guid? ResultingSaleInvoiceId { get; private set; }

    public int? Rating { get; private set; }
    public string? RatingComment { get; private set; }

    private readonly List<OrderItem> _items = new();
    public IReadOnlyCollection<OrderItem> Items => _items.AsReadOnly();

    private Order() { } // EF Core

    public Order(Guid customerId, Guid branchId, string? deliveryNote, decimal? deliveryLatitude, decimal? deliveryLongitude)
    {
        CustomerId = customerId;
        BranchId = branchId;
        DeliveryNote = deliveryNote;
        DeliveryLatitude = deliveryLatitude;
        DeliveryLongitude = deliveryLongitude;
        Status = OrderStatus.Pending;
    }

    public OrderItem AddItem(Guid productId, Guid productUnitId, decimal quantity, decimal estimatedUnitPrice)
    {
        if (Status != OrderStatus.Pending)
        {
            throw new DomainException("Cannot add items to an order that is not Pending.");
        }

        var item = new OrderItem(Id, productId, productUnitId, quantity, estimatedUnitPrice);
        _items.Add(item);
        return item;
    }

    public void Accept(Guid decidedByUserId, DateTime decidedAtUtc)
    {
        if (Status != OrderStatus.Pending)
        {
            throw new DomainException($"Cannot accept an order with status '{Status}'.");
        }

        Status = OrderStatus.Accepted;
        DecidedByUserId = decidedByUserId;
        DecidedAtUtc = decidedAtUtc;
    }

    public void Reject(Guid decidedByUserId, DateTime decidedAtUtc, string reason)
    {
        if (Status == OrderStatus.Completed || Status == OrderStatus.Rejected)
        {
            throw new DomainException($"Cannot reject an order with status '{Status}'.");
        }

        if (string.IsNullOrWhiteSpace(reason))
        {
            throw new DomainException("A rejection reason is required.");
        }

        Status = OrderStatus.Rejected;
        DecidedByUserId = decidedByUserId;
        DecidedAtUtc = decidedAtUtc;
        RejectionReason = reason;
    }

    public void Complete(Guid resultingSaleInvoiceId)
    {
        if (Status != OrderStatus.Accepted)
        {
            throw new DomainException($"Cannot complete an order with status '{Status}'.");
        }

        Status = OrderStatus.Completed;
        ResultingSaleInvoiceId = resultingSaleInvoiceId;
    }

    public void Rate(int rating, string? comment)
    {
        if (Status != OrderStatus.Completed)
        {
            throw new DomainException("Only a Completed order can be rated.");
        }

        if (Rating is not null)
        {
            throw new DomainException("This order has already been rated.");
        }

        if (rating is < 1 or > 5)
        {
            throw new DomainException("Rating must be between 1 and 5.");
        }

        Rating = rating;
        RatingComment = comment;
    }
}

public class OrderItem : Entity
{
    public Guid OrderId { get; private set; }
    public Guid ProductId { get; private set; }
    public Guid ProductUnitId { get; private set; }
    public decimal Quantity { get; private set; }

    /// <summary>سعر تقديري لحظة الطلب - عرض للزبون بس، غير مُلزِم. السعر الحقيقي يُحسَم من ProductBranch.SellingPrice لحظة Complete (نفس مبدأ CompleteSaleCommand - لا سعر من طرف العميل أبدًا).</summary>
    public decimal EstimatedUnitPrice { get; private set; }

    private OrderItem() { } // EF Core

    internal OrderItem(Guid orderId, Guid productId, Guid productUnitId, decimal quantity, decimal estimatedUnitPrice)
    {
        if (quantity <= 0)
        {
            throw new DomainException("Order item quantity must be positive.");
        }

        OrderId = orderId;
        ProductId = productId;
        ProductUnitId = productUnitId;
        Quantity = quantity;
        EstimatedUnitPrice = estimatedUnitPrice;
    }
}

File: SupermarketSystem.CustomerApp/lib/models/order.dart
/// حالات الطلب - تطابق OrderStatus بالباك إند (Domain/Ordering/Order.cs) بالضبط.
enum OrderStatus { pending, accepted, completed, rejected }

OrderStatus orderStatusFromInt(int value) {
  switch (value) {
    case 1:
      return OrderStatus.pending;
    case 2:
      return OrderStatus.accepted;
    case 3:
      return OrderStatus.completed;
    case 4:
      return OrderStatus.rejected;
    default:
      return OrderStatus.pending;
  }
}

String orderStatusLabel(OrderStatus status) {
  switch (status) {
    case OrderStatus.pending:
      return 'قيد الانتظار';
    case OrderStatus.accepted:
      return 'تم القبول - جاري التجهيز';
    case OrderStatus.completed:
      return 'تم التسليم';
    case OrderStatus.rejected:
      return 'مرفوض';
  }
}

/// يطابق OrderListItemDto (GetPendingOrders.cs المشترك مع GetCustomerOrders).
class OrderListItem {
  final String id;
  final OrderStatus status;
  final String? deliveryNote;
  final double estimatedTotal;
  final int itemsCount;
  final DateTime createdAtUtc;

  OrderListItem({
    required this.id,
    required this.status,
    required this.estimatedTotal,
    required this.itemsCount,
    required this.createdAtUtc,
    this.deliveryNote,
  });

  factory OrderListItem.fromJson(Map<String, dynamic> json) {
    return OrderListItem(
      id: json['id'] as String,
      status: orderStatusFromInt(json['status'] as int),
      deliveryNote: json['deliveryNote'] as String?,
      estimatedTotal: (json['estimatedTotal'] as num).toDouble(),
      itemsCount: json['itemCount'] as int,
      createdAtUtc: DateTime.parse(json['createdAtUtc'] as String),
    );
  }
}

class OrderItemDetail {
  final String productId;
  final String productName;
  final String unitName;
  final double quantity;
  final double estimatedUnitPrice;

  OrderItemDetail({
    required this.productId,
    required this.productName,
    required this.unitName,
    required this.quantity,
    required this.estimatedUnitPrice,
  });

  factory OrderItemDetail.fromJson(Map<String, dynamic> json) {
    return OrderItemDetail(
      productId: json['productId'] as String,
      productName: json['productName'] as String,
      unitName: json['unitName'] as String,
      quantity: (json['quantity'] as num).toDouble(),
      estimatedUnitPrice: (json['estimatedUnitPrice'] as num).toDouble(),
    );
  }
}

/// يطابق OrderDetailDto (GetOrderById/GetOrderByIdQuery.cs).
class OrderDetail {
  final String id;
  final OrderStatus status;
  final String? deliveryNote;
  final String? rejectionReason;
  final List<OrderItemDetail> items;
  final DateTime createdAtUtc;
  final int? rating;
  final String? ratingComment;

  OrderDetail({
    required this.id,
    required this.status,
    required this.items,
    required this.createdAtUtc,
    this.deliveryNote,
    this.rejectionReason,
    this.rating,
    this.ratingComment,
  });

  factory OrderDetail.fromJson(Map<String, dynamic> json) {
    return OrderDetail(
      id: json['id'] as String,
      status: orderStatusFromInt(json['status'] as int),
      deliveryNote: json['deliveryNote'] as String?,
      rejectionReason: json['rejectionReason'] as String?,
      items: (json['items'] as List<dynamic>)
          .map((e) => OrderItemDetail.fromJson(e as Map<String, dynamic>))
          .toList(),
      createdAtUtc: DateTime.parse(json['createdAtUtc'] as String),
      rating: json['rating'] as int?,
      ratingComment: json['ratingComment'] as String?,
    );
  }
}
